driver-core/src/main/com/mongodb/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -0,0 +1,193 @@
+
+    AsyncChangeStreamBatchCursor(final ChangeStreamOperation<T> changeStreamOperation,
+                                 final AsyncReadBinding binding,
+                                 final AsyncQueryBatchCursor<RawBsonDocument> wrapped) {
+        this.changeStreamOperation = changeStreamOperation;
+        this.resumeToken = changeStreamOperation.getResumeToken();
+        this.wrapped = wrapped;
+        this.binding = binding;
+    }
+
+    AsyncQueryBatchCursor<RawBsonDocument> getWrapped() {
+        return wrapped;
+    }
+
+    ServerCursor getServerCursor() {
+        return wrapped.getServerCursor();
+    }
+
+    @Override
+    public void next(final SingleResultCallback<List<T>> callback) {
+        resumeAbleOperation(new AsyncBlock() {
+            @Override
+            public void apply(final AsyncQueryBatchCursor<RawBsonDocument> cursor,
+                              final SingleResultCallback<List<RawBsonDocument>> callback) {
+                cursor.next(callback);
+            }
+        }, convertResultsCallback(callback));
+    }
+
+    @Override
+    public void tryNext(final SingleResultCallback<List<T>> callback) {
+        resumeAbleOperation(new AsyncBlock() {
+            @Override
+            public void apply(final AsyncQueryBatchCursor<RawBsonDocument> cursor,
+                              final SingleResultCallback<List<RawBsonDocument>> callback) {
+                cursor.tryNext(callback);
+            }
+        }, convertResultsCallback(callback));
+    }
+
+    @Override
+    public void setBatchSize(final int batchSize) {
+        wrapped.setBatchSize(batchSize);
+    }
+
+    @Override
+    public int getBatchSize() {
+        return wrapped.getBatchSize();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return wrapped.isClosed();
+    }
+
+    @Override
+    public void close() {
+        wrapped.close();
+    }
+
+    private interface AsyncBlock {
+        void apply(AsyncQueryBatchCursor<RawBsonDocument> cursor, SingleResultCallback<List<RawBsonDocument>> callback);
+    }
+
+    private void resumeAbleOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback) {
+        boolean msgInFlight = inFlight.getAndSet(true);
+        if (msgInFlight) {
+            callback.onResult(null,
+                    new MongoChangeStreamException("Another asynchronous operation on this Changes Stream is already in progress"));
+            return;
+        }
+
+        asyncBlock.apply(wrapped, new SingleResultCallback<List<RawBsonDocument>>() {
+            @Override
+            public void onResult(final List<RawBsonDocument> result, final Throwable t) {
+                inFlight.set(false);
+                if (t == null || !changeStreamOperation.isResumable()) {
+                    callback.onResult(result, t);
+                } else if (t instanceof MongoServerException && !(t instanceof MongoNotPrimaryException
```

Comment:
See my comment in ChangeStreamOperation.  I think a retriable error is one of 

* MongoNotPrimaryException
* MongoQueryException 
* MongoSocketException

All others can be passed through